WIP DYN-2538 Generate single RenderPackage per Node

saintentropy opened the pull request:
### Purpose

Note: I was cleaning up my Dynamo git stash over the holiday break and this was too good to leave on the shelf.  This builds on some work that @LongNguyenP, @Dewb, @aparajit-pratap, and I looked at a while back.  I picked up and fixed the regressions we noted previously. 

The purpose of this PR is a performance and memory optimization for the render pipeline.  The main architectural change is modifying the assumption that we generate a single `RenderPackage` object per every individual return object for a node. Instead we aggregate all node tessellation data into a single `RenderPackage` recursively.  This improves tessellation performance as the `HelixRenderPackage` is a large .NET object with a non-trivial init time in addition to a large saving in memory; specifically one render package with all the tessellation data is much smaller then many render packages with partial tessellation data.  This approach also creates performance benefit and memory savings when the tessellated data is added to the Helix scene.  Processing one RenderPackage per `AggregateRenderPackages` call is dramatically more efficient than numerous render packages as there are numerous collection interactions and temporary allocation.  In code, this change was enabled by generating the RenderPackage first in `GetRenderPackagesFromMirrorData` before the recursive loop to fill the render package data.  Previously it was generated within the loop.

There are a few things that require changes to support features which assumed a single item = a single RenderPackage.

* Labels (ie the labels that appear in the scene context by selecting items in the node preview).  The existing implementation encoded label information in the `Description` property of the `RenderPackage`.  That encoding was not well documented making it nearly impossible to add labels to the your own tessellation method overrides.  Here I introduce a new interface, `IRenderLabels`, which provides methods on the render package to add label data and placement.  The new underlying storage implementation matches the existing `LabelPlaces` data structure in the `HelixWatch3DViewModel` so that adding labels requires much less string parsing and processing. 

* Texture maps.  The original texture map implementation in the `Colors` and `ColorStride` property assumed only one texture map per the mesh geometry in the `RenderPackage`.  To support aggregation of multiple mesh geometries within a single `RenderPackage` with unique textures, I introduce a new interface, `IRenderPackageSupplement`, which supports adding lists of Colors and ColorStrides along with their associated mesh vertex ranges.  The `Colors` and `ColorStride` property as well as the `SetColors` method are marked as obsolete with notes to use the new methods in the `IRenderPackageSupplement`.

* Color adding methods on `IRenderPackage`.  While methods to add vertex color like  `AddPointVertexColor` still work fine with this change, when utilized in tessellation calls, methods like `ApplyPointVertexColors` overwrite the previous data vs appending.  That can cause data loss as they are currently used.  I introduce new methods to `IRenderPackageSupplement` like `InsertPointVertexColorRange` and `AppendPointVertexColorRange` to allow additions without overwriting existing data.  The old methods are also marked as obsolete with notes to use the new methods in the `IRenderPackageSupplement`.  The GeometryColor nodes are updated to use the new methods.

These new features are being introduced with a minor release and not Dynamo 3.0.  While we can update the tessellation implementation in Dynamo and Dynamo integrations, Third Party tessellation functions also need to stay working.  For each of the Obsolete methods above I introduced an Exception to allow the `GetRenderPackagesFromMirrorData` implementation to detect their usage.  Specifically within `IRenderPackageSupplement` there is a boolean `AllowLegacyColorOperations` that can be set to false which will cause the `LegacyRenderPackageMethodException` to be thrown when the obsolete methods are called.  In that case, we roll back any changes to the `RenderPackage` that occurred within the `Tesselate` call and generate a new `RenderPackage` with `AllowLegacyColorOperations` set to true and re-Tesselate.  This individual render package is added to the RenderPackageCache and the outer loop resumes.  In the transition to Dynamo 3.0, the obsolete methods can be removed as well as this fallback.

Todo
* Add tests for the new functionality
* Update Tessellate implementation in DynamoRevit, other integrations, samples, or internal packages.  This is not strictly required due to the fallback mechanism. 
* Fix the Label.ByPointAndString node to work with the new label mechanism so that the node actually work.
* ITransformable needs to be refactored similarly to the Texture Map implementation.  This may be covered under a separate PR as we can possibly take advantage of the refactor to introduce instancing mechanism. 

Examples

https://user-images.githubusercontent.com/2145751/104148265-98006400-539f-11eb-980e-595a01f6b856.mp4

https://user-images.githubusercontent.com/2145751/104148270-9df64500-539f-11eb-9c8f-20ed9d0c27e5.mp4

https://user-images.githubusercontent.com/2145751/104148276-a2baf900-539f-11eb-9edf-a803ab32acc3.mp4

https://user-images.githubusercontent.com/2145751/104148280-a64e8000-539f-11eb-997d-f8235882d69a.mp4

For most point and curve tessellation the speed and memory improvement will be linear based on the number of tessellated items.  However for meshes the speed and memory improvement is dramatic. 

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers

@mjkkirschner @aparajit-pratap @QilongTang

mjkkirschner commented:
@aparajit-pratap one downside of marking this PR as a draft is that the tests do not run - we'd have to modify the code that runs on the self serve pull job - and I think it makes use of https://github.com/microsoft/PowerShellForGitHub ... I gave up once I got that far.

mjkkirschner commented:
@saintentropy it would be good to run the wpf visualization test jobs - there are versions of that job that run with warp and another that runs on a real GPU.

aparajit-pratap commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -212,104 +280,173 @@ protected override void HandleTaskExecutionCore()
+            var nEnd = plane.Origin.Add(plane.Normal.Scale(2.5));
+            package.AddLineStripVertex(nEnd.X, nEnd.Y, nEnd.Z);
+
+            for (var i = 0; i < 5; i++)
```
An explanation of why this is 5 would help.

saintentropy replied:
Yep that would be good.  5 lines

aparajit-pratap commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -212,104 +280,173 @@ protected override void HandleTaskExecutionCore()
-                        {
-                            package.AddTriangleVertexNormal(plane.Normal.X, plane.Normal.Y, plane.Normal.Z);
-                        }
+            for (var i = 0; i < 6; i++)
```
An explanation of why this is 6?

saintentropy replied:
Yes I can... 6 mesh vertex

saintentropy replied:
Also we can move this into LibG too.  Not sure why it lives here.

saintentropy commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -149,54 +149,122 @@ protected override void HandleTaskExecutionCore()
-                    //If the package has a transform that is not the identity matrix
-                    //then set requiresCustomTransform to true.
-                    if (packageWithTransform != null && packageWithTransform.Transform.SequenceEqual(
-                        new double[] { 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1 }) == false)
```
I think that case is valid.... I just don't think that is what our implementation supports currently.  Today you add your tessellated geometry and then you can additionally add a transform.  That transform does not support instancing it only transforms all the geometry in the render package.  I think the original use case being able to tesselate an object once and then move it around.

saintentropy replied:
I think instancing would be much more interesting and we should add that as an enhancement.  I think we should do it as a follow up PR.  It would need a similar data structure to the color aspects where you can apply numerous transforms to any range of geometry data.  The advantage would be that we could and tessellated geometry data once and supply a transform matrix for each instance.  There would also need to changes to the helix implementation to support real instancing

saintentropy commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -149,54 +149,122 @@ protected override void HandleTaskExecutionCore()
+                    catch (LegacyRenderPackageMethodException)
                     {
-                        (packageWithTransform).RequiresCustomTransform = true;
+                        //At this point we have detected an implementation of Tessellate which is using legacy color methods
```
We do not use the API's that throw `LegacyRenderPackageMethodException` in LibG so no changes are required.  Only some of the Color methods fall in this category.  `void ApplyPointVertexColors(byte[] colors)` for example as it replaces all the vertex colors in the render packages.  This would overwrite any color data that had previously been added.  `void AddPointVertexColor(byte red, byte green, byte blue, byte alpha)` is ok as it just continues to append data.

aparajit-pratap commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -212,104 +280,173 @@ protected override void HandleTaskExecutionCore()
+        }
+
+        private static void LegacyPackageEnsureColorExistsPerVertex(IRenderPackage pkg, int previousPointVertexCount, int previousLineVertexCount, int previousMeshVertexCount)
+        {
```
```
var colorCount = pkg.PointVertexColors.Count() / 4;
if (pkg.PointVertexCount > previousPointVertexCount
                && colorCount > pkg.PointVertexCount) 
{
    for (var i = colorCount; i < pkg.PointVertexCount; i++)
    { ...
```

saintentropy replied:
This would be more clear...  Since this is on the hot path and the `PointVertexColors` is and IEnumerable<byte> I was a bit allergic to calling `Count()` unless I really had to.  Let me do some more reading on the performance impact

aparajit-pratap replied:
My suggestion is to reuse the count instead of calling `Count()` on the innumerable in multiple places.

aparajit-pratap commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -212,104 +280,173 @@ protected override void HandleTaskExecutionCore()
+                var count = package.PointVertexCount - packageSupplement.PointVertexColorCount;
+                if (count > 0)
+                {
+                    packageSupplement.AppendPointVertexColorRange(CreateColorByteArrayOfSize(count, DefR, DefG, DefB, DefA));
```
How is this different from `AddPointVertexColor` called in a for loop?

saintentropy replied:
functionally there isn't a huge difference... under the hood it uses `AddRange`.  I think it just is clearer code.

aparajit-pratap commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -212,104 +280,173 @@ protected override void HandleTaskExecutionCore()
+            }
+        }
+
+        private static void LegacyPackageEnsureColorExistsPerVertex(IRenderPackage pkg, int previousPointVertexCount, int previousLineVertexCount, int previousMeshVertexCount)
```
So this method essentially does the same thing as `EnsureColorExistsPerVertex` but with the old APIs? I suppose this check is needed only now after the change to a single render package per node? If so, when will there be a case where the package will not be an `IRenderPackageSupplement` but there would still be a single package per node?

aparajit-pratap replied:
Is this method needed for custom third party `IRenderPackage `types that do not implement `IRenderPackageSupplement` but will still be running Dynamo after these changes?

saintentropy replied:
Yes this is for third party implementation of `IRenderPackage`.  Our Helix and Default implementations in Dynamo  would add `IRenderPackageSuppliment`.  The color check is done regardless of render package per `IGraphicItem` or per node.  This was previously handled here -> https://github.com/DynamoDS/Dynamo/blob/master/src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs#L290.  This just ensures color info is added to the RenderPackage before being passed to the Helix View

aparajit-pratap commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -212,104 +280,173 @@ protected override void HandleTaskExecutionCore()
+                && pkg.LineStripVertexColors.Any()
+                && pkg.LineStripVertexColors.Count() / 4 > pkg.LineVertexCount)
+            {
+                for (var i = pkg.LineStripVertexColors.Count() / 4; i < pkg.LineVertexCount; i++)
```
Similar refactor as above will be useful.

aparajit-pratap commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -212,104 +280,173 @@ protected override void HandleTaskExecutionCore()
+                && pkg.MeshVertexColors.Any()
+                && pkg.MeshVertexColors.Count() / 4 > pkg.MeshVertexCount)
+            {
+                for (var i = pkg.MeshVertexColors.Count() / 4; i < pkg.MeshVertexCount; i++)
```
Similar refactor as above will be useful.

aparajit-pratap commented on src/DynamoCoreWpf/Rendering/HelixRenderPackage.cs:
```diff
@@ -555,12 +521,281 @@ public IEnumerable<int> PointIndices
+        /// <summary>
+        /// The number of point vertices colors in the package.
+        /// </summary>
+        public int PointVertexColorCount => points.Colors.Count;
```
Why do these implementations differ from those in `DefaultRenderPackage`?

saintentropy replied:
Because the helix render package is highly optimized to store data in Helix data structure.  Color, Vertex, etc...   This makes for a faster transfer to the View

aparajit-pratap commented on src/VisualizationTests/HelixWatch3DViewModelTests.cs:
```diff
@@ -1189,7 +1189,7 @@ private Watch3DView FindFirstWatch3DNodeView()
         }
 
-        [Test]
+        [Test, Category("Failure")]
```
Why is this failing now?

saintentropy replied:
The two meshes do not exists any more within the `Element3DDictionary`.   Previously this RenderPackageCache would have added `var_88f09e6c057f4a5c95a7f4f0b25161e2:0:0:0":mesh` and `var_88f09e6c057f4a5c95a7f4f0b25161e2:0:0:1:mesh`.  Now as far as the Helix view is concerned there is only one mesh object related to this node identifier in the `Element3DDictionary`.  That is `var_88f09e6c057f4a5c95a7f4f0b25161e2:mesh` I think.  The indexes are no longer added here as this new implementation only adds one bucket for mesh, curve, and vertex data to the Helix dictionary per node.  In fact that is why this is so much faster.  1 point bucket vs potentially 100K is much much faster.

saintentropy replied:
And much more memory efficient.

aparajit-pratap replied:
Why not remove this test in that case?

saintentropy replied:
That is good.  I just marked them as fail initially as to clearly communicate behavioral changes

saintentropy commented on src/Libraries/GeometryColor/GeometryColor.cs:
```diff
@@ -241,7 +241,9 @@ private static Color[][] ComputeColorMap(Surface surface, IEnumerable<UV> uvs, C
         private void CreateColorMapOnSurface(Color[][] colorMap , IRenderPackage package, TessellationParameters parameters)
```
👍

saintentropy commented on src/NodeServices/GraphicsDataInterfaces.cs:
```diff
@@ -196,6 +202,124 @@ public interface IRenderPackage
+        /// <summary>
+        /// A flag indicating whether the render package should auto generate labels
+        /// </summary>
+        bool AutoGenerateLabels { get; set; }
```
This is the classic replication index label that is used to indicate the position of the item when you select it in the preview list.

saintentropy commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -212,104 +280,173 @@ protected override void HandleTaskExecutionCore()
+            if (package.PointVertexCount > previousPointVertexCount)
+            {
+                var count = package.PointVertexCount - packageSupplement.PointVertexColorCount;
+                if (count > 0)
```
The count in 391 is to check if there has been a change to the geometry data in the render package when the tessellation method was called.  It could be that this specific tessellation object did not have any point vertex data and then no action is needed.  The second check on line 393 is verify that within the package the color and point vertex information is synchronized.  If the the tessellation method added point vertex data but did not add color, this method would add default colors for the new geometry data.

saintentropy commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -212,104 +280,173 @@ protected override void HandleTaskExecutionCore()
+            if (package.LineVertexCount > previousLineVertexCount)
+            {
+                var count = package.LineVertexCount - packageSupplement.LineVertexColorCount;
+                if (count > 0)
```
same situation as point vertex.  Default colors have to be added if tessellation does not add color info.   This was actually also true in before the refactor.

saintentropy replied:
Current implementation is here -> https://github.com/DynamoDS/Dynamo/blob/master/src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs#L290

aparajit-pratap commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -149,54 +149,122 @@ protected override void HandleTaskExecutionCore()
-                        (packageWithTransform).RequiresCustomTransform = true;
+                        //At this point we have detected an implementation of Tessellate which is using legacy color methods
+                        //We roll back the primary renderPackage to it previous state before calling tessellation again.
+                        package = RollBackPackage(package, previousPointVertexCount, previousLineVertexCount,
```
I still don't understand why this needs to be rolled back if it's going to be discarded anyway when we return in line 254. I don't see `package `is being used anywhere beyond this point inside this condition?

aparajit-pratap commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -212,104 +280,173 @@ protected override void HandleTaskExecutionCore()
+            var b = Point.ByCartesianCoordinates(cs, -s, s, 0);
+            var c = Point.ByCartesianCoordinates(cs, -s, -s, 0);
+            var d = Point.ByCartesianCoordinates(cs, s, -s, 0);
+            //Todo Dispose cs, a, b, c, d?
```
Can we add these Dispose statements in this PR while we are at it?

aparajit-pratap commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -212,104 +280,173 @@ protected override void HandleTaskExecutionCore()
+        {
+            if (pkg.PointVertexCount > previousPointVertexCount
+                && pkg.PointVertexColors.Any()
+                && pkg.PointVertexColors.Count() / 4 > pkg.PointVertexCount)
```
Shouldn't this check be `pkg.PointVertexCount > pkg.PointVertexColors.Count() / 4`

aparajit-pratap commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -212,104 +280,173 @@ protected override void HandleTaskExecutionCore()
-                package.IsSelected = isNodeSelected;
+            if (pkg.LineVertexCount > previousLineVertexCount
+                && pkg.LineStripVertexColors.Any()
+                && pkg.LineStripVertexColors.Count() / 4 > pkg.LineVertexCount)
```
Shouldn't this check be` pkg.LineVertexCount > pkg.LineVertexColors.Count() / 4`?

aparajit-pratap commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -212,104 +280,173 @@ protected override void HandleTaskExecutionCore()
-                renderPackageCache.Add(package, outputPortId);
+            if (pkg.MeshVertexCount > previousMeshVertexCount
+                && pkg.MeshVertexColors.Any()
+                && pkg.MeshVertexColors.Count() / 4 > pkg.MeshVertexCount)
```
Shouldn't this check be `pkg.MeshVertexCount > pkg.MeshVertexColors.Count() / 4`?

aparajit-pratap commented on src/DynamoCoreWpf/Rendering/HelixRenderPackage.cs:
```diff
@@ -555,12 +521,281 @@ public IEnumerable<int> PointIndices
+            var message = string.Empty;
+            if (!HasValidStartEnd(startIndex, endIndex, points, out message))
+            {
+                if (!string.IsNullOrEmpty(message))
+                {
+                    throw new Exception(message);
+                }
```
Why throw an exception only if there is a message? In any case I think the message existing and `HasValidStartEnd` returning false go hand-in-hand.

aparajit-pratap commented on src/DynamoCoreWpf/Rendering/HelixRenderPackage.cs:
```diff
@@ -555,12 +521,281 @@ public IEnumerable<int> PointIndices
+        ///// <summary>
+        ///// Get label data; label string and associated position
+        ///// </summary>
+        public List<Tuple<string, float[]>> LabelData
```
Can we just have one property - either `LabelPlaces` or `LabelData` instead of having both? I think they are essentially the same data, just formatted differently, I'm I'm not wrong?

aparajit-pratap commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -149,54 +149,122 @@ protected override void HandleTaskExecutionCore()
+            }
+        }
+
+        private void GetRenderPackageFromGraphicItem(Guid outputPortId, IGraphicItem graphicItem, IRenderPackage package, string labelKey)
```
Are you going to rename this to something more appropriate like you discussed earlier?

aparajit-pratap commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -149,54 +149,121 @@ protected override void HandleTaskExecutionCore()
+            }
+        }
+
+        private void GetTessellationDataFromGraphicItem(Guid outputPortId, IGraphicItem graphicItem, IRenderPackage package, string labelKey)
```
I think `package `should be passed as a `ref` parameter as it can be assigned a new object?

saintentropy replied:
Definitely!  Good catch

aparajit-pratap commented on src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs:
```diff
@@ -364,13 +363,20 @@ private static void CreatePlaneTessellation(IRenderPackage package, Plane plane)
             }
+
+            //dispose helper geometry
+            a.Dispose();
+            b.Dispose();
+            c.Dispose();
+            d.Dispose();
+            cs.Dispose();
```
How about using `using` statements instead?

saintentropy replied:
I tried this and it was more convoluted then the dispose.  I think just keep it simple here.

aparajit-pratap commented:
@saintentropy thanks for addressing the review comments and adding inline documentation. Also thanks for clearer names for the new APIs. I have just two remaining questions, then LGTM.

saintentropy commented:
Hi @aparajit-pratap Added some of fixes from the comments and a lot more tests

saintentropy commented:
I also added the single item label fix.  The regression was introduced in this PR -> https://github.com/DynamoDS/Dynamo/pull/10989.  That PR also added the non-functional DisplayLabel node back into core so this PR fixes both the node and the label regression.

aparajit-pratap commented on src/VisualizationTests/HelixRenderPackageTests.cs:
```diff
@@ -157,6 +158,398 @@ public void HelixRenderPackage_MeshColorBufferCheck()
+
+            var testColors = WhiteByteArrayOfLength(5);
+
+            try
+            {
+                p.AppendMeshVertexColorRange(testColors);
+            }
+            catch(Exception e)
+            {
+                Assert.AreEqual(e.GetType(), typeof(Exception));
+            }
```
`Assert.ThrowsException<Exception>(() => p.AppendMeshVertexColorRange(testColors));`

aparajit-pratap replied:
How are these exceptions dealt with in Dynamo in a real situation?

saintentropy replied:
All the exceptions (my new ones and the old ones) within the `RenderPacakge` implementation are caught here -> https://github.com/DynamoDS/Dynamo/blob/master/src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs#L303.  The net affect is that the node does not display geometry in the background preview.

saintentropy replied:
The only difference in the new implementation is the specific handling of the `LegacyRenderPackageMethodException

aparajit-pratap commented on src/VisualizationTests/HelixRenderPackageTests.cs:
```diff
@@ -157,6 +158,398 @@ public void HelixRenderPackage_MeshColorBufferCheck()
+            {
+                p.AppendLineVertexColorRange(testColors);
+            }
+            catch (Exception e)
```
Can we make these exceptions thrown be more specific?

saintentropy replied:
I think I can cover these in the next PR as I add more test coverage.
